### PR TITLE
Add basic vllm support

### PR DESCRIPTION
**Pull request by ericcurtin**
This will only work with --nocontainer for now. Requires vllm > 6.0 .

### CODE REVIEW

**Comment by ericcurtin** [comment]
I think we make a breaking change here, "vllm serve" only serves .gguf file correctly with .gguf extension, so we might be better off storing files like:

~/.local/share/ramalama/models/ollama/granite-code:latest

like this instead:

~/.local/share/ramalama/models/ollama/granite-code:latest.gguf

**Comment by rhatdan** [comment]
Can we play games with symbolic links?  I am not that concerned about breaking changes at this point, since we have not released anything yet.

**Comment by ericcurtin** [comment]
Above are symlinks, so the cleanest way is to make a breaking change and change this symlink:

~/.local/share/ramalama/models/ollama/granite-code:latest

to this one:

~/.local/share/ramalama/models/ollama/granite-code:latest.gguf

Also logged a bug with vllm:

https://github.com/vllm-project/vllm/issues/7993

It's supposed to be possible to do this without the extension by using flags like "--quantization gguf --load-format gguf" but it doesn't work. GGUF inference is like 3 weeks old in vllm, they were using other formats before.

**Comment by ericcurtin** [comment]
ramalama would solve this problem for many:

https://github.com/vllm-project/vllm/issues/1441

This is one of the most commented issues in vllm... vllm focuses more on enterprise grade GPUs, so it frustrates people they don't support Apple Devices...

Whereas with ramalama, one could test locally via:

ramalama --runtime llama.cpp serve, etc.

and if they want to use vllm on a deployment server, they could do:

ramalama --runtime vllm serve, etc.

potentially....

**Comment by ericcurtin** [comment]
Some of these things are bug fixes I'm gonna split those out into another PR

**Comment by kannon92** [comment]
Excited for this! https://github.com/ollama/ollama/issues/3953

There was some interest in that issue in the feature so we should make sure to highlight it as a feature.

**Comment by ericcurtin** [comment]
vllm support will probably be completed, once there is a new release of vllm, to get some PRs like:

https://github.com/vllm-project/vllm/pull/8056

vllm release pretty frequently though, so it shouldn't be too far away...

**Comment by matbee-eth** [comment]
> I think we may need to make a breaking change here, "vllm serve" only serves .gguf file correctly with .gguf extension, so we might be better off storing files like:
> 
> 
> 
> ~/.local/share/ramalama/models/ollama/granite-code:latest
> 
> 
> 
> like this instead:
> 
> 
> 
> ~/.local/share/ramalama/models/ollama/granite-code:latest.gguf
> 
> 

I'm not sure how feasible it is for this project, but a FuSE filesystem for dynamically managing models instead of symlinks could be an option

**Comment by ericcurtin** [comment]
> > I think we may need to make a breaking change here, "vllm serve" only serves .gguf file correctly with .gguf extension, so we might be better off storing files like:
> > ~/.local/share/ramalama/models/ollama/granite-code:latest
> > like this instead:
> > ~/.local/share/ramalama/models/ollama/granite-code:latest.gguf
> 
> I'm not sure how feasible it is for this project, but a FuSE filesystem for dynamically managing models instead of symlinks could be an option

Not sure we wanna go here right now. For one, we support macOS native. This problem in particular will go away in the next version of vllm when the above PR gets merged.

But... I'd still be interested in hearing a more detailed example/explanation of why FuSE filesystem would help us here...

Although it would be preferably to have a fs-agnostic solution if at all possible...

**Comment by ericcurtin** [comment]
This is very basic support for the --nocontainer case, but may unblock other efforts like:

#150

File: ramalama/cli.py [hunk 1]
```diff
@@ -38,6 +38,12 @@ def init_cli():
     )
     parser.add_argument("--store", default=get_store(), help="store AI Models in the specified directory")
     parser.add_argument("--dryrun", action="store_true", help="show container runtime command without executing it")
+    parser.add_argument(
+        "--runtime",
+        default="llama.cpp",
+        choices=["llama.cpp", "vllm"],
+        help="Specify the runtime to use, valid options are 'llama.cpp' and 'vllm'",
```
Comment (rhatdan) [inline ramalama/cli.py:45]: We are defaulting to lowercase help messages now.

Comment (rhatdan) [reply ramalama/cli.py:45]: Could you also fix:

    parser.add_argument("ARGS", nargs="*", help="Additional options to pass to the AI Model")


While you are in there?

Comment (ericcurtin) [reply ramalama/cli.py]: Done

**Comment by rhatdan** [comment]
LGTM

**Comment by ericcurtin** [comment]
@kannon92 basic support for this was merged

**Comment by cdoern** [comment]
@ericcurtin from an ilab side of things I have been trying to see if we can get operations like generate on a mac using vllm. Could this help with that? it seems like the discourse here is saying that vllm works on the mac using ramalama?

**Comment by ericcurtin** [comment]
What this PR enables is the option to do:

```
$ ramalama --runtime vllm serve some-model
```

or

```
$ ramalama --runtime llama.cpp serve some-model
```

llama.cpp and vllm have different primary design goals. llama.cpp was written primarily with commodity grade systems in mind like a macbook, mac mini, etc. vllm was written with almost the opposite priority, to run on huge enterprise-grade systems with lots of vram and powerful gpus.

So with this change, one could test on macOS with llama.cpp as a runtime but deploy to some enterprise server with vllm as a runtime.

**Pull request merged**